File: merger_engine.py
import requests
import re
from datetime import datetime

# List of sources provided by user
SOURCES = [
    "https://adaway.org/hosts.txt",
    "https_hagezi_popupads", # Placeholder for logic below
]

def fetch_urls():
    """The actual list from your request."""
    return [
        "https://adaway.org/hosts.txt",
        "https://cdn.jsdelivr.net/gh/hagezi/dns-blocklists@latest/adblock/popupads.txt",
        "https://cdn.jsdelivr.net/gh/hagezi/dns-blocklists@latest/hosts/ultimate.txt",
        "https://cdn.jsdelivr.net/gh/hagezi/dns-blocklists@latest/hosts/ultimate-compressed.txt",
        "https://big.oisd.nl",
        "https://raw.githubusercontent.com/r0xd4n3t/pihole-adblock-lists/main/pihole_adlists.txt",
        "https://v.firebog.net/hosts/Easylist.txt",
        "https://raw.githubusercontent.com/StevenBlack/hosts/master/data/StevenBlack/hosts",
        "https://pgl.sysdin.de/hosts/serverlist_php?hostformat=hosts&mimetype=plaintext&useip=0.0.0.0", # Corrected URL structure for parsing
        "https://raw.githubusercontent.com/FiltersHeroes/KADhosts/master/KADhosts.txt"
    ]

def parse_line(line):
    """Extracts host from common adblock formats (e.g., 127.0.0.1 domain.com or domain.com)"""
    line = line.strip()
    if not line or line.startswith("#") or "://" in line: # Skip comments and full URLs if they aren't host entries
        return None
    # Handle 0.0.0.0/127.0.0.1 prefix common in hosts files
    parts = re.split(r'\s+', line)
    host_entry = parts[0] if len(parts) > 1 else parts[0]
    return host_entry

def merge_lists():
    all_entries = set() # Using a Set to automatically handle duplicates (Deduplication/Factorization step)
    urls = fetch_urls()
    processed_count = 0
    errors = []

    print(f"--- Starting Merge Process: {datetime.now()} ---")

    for url in urls:
        try:
            response = requests.get(url, timeout=15)
            if response.status_code == 200:
                lines = response.text.split('\n')
                count_added = 0
                for line in lines:
                    entry = parse_line(line)
                    if entry and entry not in all_entries:
                        all_entries.add(entry)
                        count_added += 1
                print(f"[SUCCESS] {url}: Added/Verified {count_added} new entries.")
            else:
                errors.append(f"Failed to fetch {url} (Status: {response.status_code})")
        except Exception as e:
            errors.append(f"Error processing {url}: {str(e)}")

    # Write the final clean list back to AdBlock_Full_List.txt in our repo folder structure
    output_path = "AdBlock_Full_List.txt" 
    with open(output_path, 'w') as f:
        for entry in sorted(list(all_entries)): # Sorting ensures a deterministic output for Git diffs
            f.write(entry + "\n")

    print(f"\n--- Final Report ---")
    print(f"Total unique entries preserved after deduplication: {len(all_entries)}")
    if errors:
        print("Errors encountered:")
        for err in errors: print(f"- {err}")
    else:
        print("All sources processed successfully.")

if __name__ == "__main__":
    merge_lists()

File: merger_engine.js
/**
 * @file merger_engine.js
 * @description Advanced Ad-Block Merger Engine v2.0
 * Features: Modular architecture, Error-resilient fetching, and Auto-Deduplication.
 * @author OpenClaw Assistant
 */

const https = require('https');
const fs = require('fs');
const path = require('path');

/**
 * CONFIGURATION
 * Add or remove blocklist sources here.
 */
const SOURCES = [
    { name: "AdAway", url: "https://adaway.org/hosts.txt", category: "General" },
    { name: "Hagezi-Popup", url: "https://cdn.jsdelivr.net/gh/hagezi/dns-blocklists@latest/adblock/popupads.txt", category: "Popups" },
    { name: "Hagezi-Ultimate", url: "https://cdn.jsdelivr.net/gh/hagezi/dns-blocklists@latest/hosts/ultimate.txt", category: "Privacy" },
    { name: "OISD-Big", url: "https://big.oisd.nl", category: "Ads" },
    { name: "PiHole-r0xd4n3t", url: "https://raw.githubusercontent.com/r0xd4n3t/pihole-adblock-lists/main/pihole_adlists.txt", category: "Ads" },
    { name: "Firebog-Easylist", url: "https://v.firebog.net/hosts/Easylist.txt", category: "General" },
    { name: "StevenBlack", url: "https://raw.githubusercontent.com/StevenBlack/hosts/master/data/StevenBlack/hosts", category: "Privacy" },
    { name: "Yoyo", url: "https://pgl.yoyo.org/adservers/serverlist.php?hostformat=hosts&mimetype=plaintext&useip=0.0.0.0", category: "Ads" },
    { name: "KADhosts", url: "https://raw.githubusercontent.com/FiltersHeroes/KADhosts/master/KADhosts.txt", category: "Regional" }
];

const OUTPUT_FILE = path.join(__dirname, 'AdBlock_Full_List.txt');

/**
 * MODULE: Fetcher
 * @param {string} url - The URL to download
 * @returns {Promise<string>} The raw text content
 */
async function fetchSource(url) {
    return new Promise((resolve, reject) => {
        https.get(url, { timeout: 10000 }, (res) => {
            if (res.statusCode < 200 || res.statusCode >= 300) {
                return reject(new Error(`HTTP ${res.statusCode}`));
            }
            let body = '';
            res.on('data', chunk => body += chunk);
            res.on('end', () => resolve(body));
        }).on('error', reject);
    });
}

/**
 * MODULE: Parser
 * Extracts valid hostnames from raw data, cleaning comments and whitespace.
 * @param {string} data - Raw text from a source
 * @returns {string[]} Array of cleaned hostnames
 */
function parseData(data) {
    return data.split('\n')
        .map(line => line.trim())
        .filter(line => line && !line.startsWith('#')) // Remove comments
        .map(line => {
            const parts = line.split(/\s+/);
            // Handle both "0.0.0.0 domain.com" and "domain.com" formats
            return parts.length > 1 ? parts[1] : parts[0];
        })
        .filter(entry => entry && entry.includes('.') && !entry.includes('://')); // Basic domain validation
}

/**
 * MAIN EXECUTION LOOP
 */
async function runEngine() {
    console.log(`🚀 Starting Modular Merge Engine at ${new Date().toLocaleString()}`);
    const MasterSet = new Set();
    let totalStats = { success: 0, failed: 0 };

    for (const source of SOURCES) {
        try {
            process.stdout.write(`📥 Fetching ${source.name}... `);
            const raw = await fetchSource(source.url);
            const cleaned = parseData(raw);
            
            cleaned.forEach(item => MasterSet.add(item));
            console.log(`Done! (+${cleaned.length} items)`);
            totalStats.success++;
        } catch (err) {
            console.log(`❌ FAILED: ${err.message}`);
            totalStats.failed++;
        }
    }

    // Sort alphabetically for clean Git diffs
    const sortedResult = Array.from(MasterSet).sort();
    
    // Header for the output file
    const header = `# AdBlock Full List
# Generated: ${new Date().toISOString()}
# Total Unique Entries: ${MasterSet.size}
# Sources Processed: ${totalStats.success}/${SOURCES.length}
# ------------------------------------------------------------\n`;

    fs.writeFileSync(OUTPUT_FILE, header + sortedResult.join('\n') + '\n');

    console.log(`\n✅ MERGE COMPLETE`);
    console.log(`📊 Total Unique Domains: ${MasterSet.size}`);
    console.log(`📂 Output saved to: ${OUTPUT_FILE}`);
}

runEngine();
